packages/compiler-core/src/transforms/vOn.ts

```diff
@@ -28,6 +30,51 @@ export interface VOnDirectiveNode extends DirectiveNode {
   exp: SimpleExpressionNode | undefined
 }
 
+export const unfoldExpression = (
+  exp: CompoundExpressionNode
+): SimpleExpressionNode => {
+  let loc: SourceLocation = {
+    source: '',
+    start: { line: 1, column: 1, offset: 0 },
+    end: { line: 1, column: 1, offset: 0 }
+  }
+  let content = ''
+
+  for (let i = 0; i < exp.children!.length; i++) {
+    let node = exp.children[i]
+    if (isString(node)) {
+      loc.source += node
+      content += node
+    } else {
+      let simpleExp = node as SimpleExpressionNode
+      !i && (loc.start = simpleExp.loc.start)
+      loc.end = simpleExp.loc.end
+      loc.source += simpleExp.loc.source
+      content += simpleExp.content
+    }
+  }
+
+  return createSimpleExpression(content, false, loc)
+}
+
+// Event expression wrap to make sure it is a function.
+// See issue #2605.
+export const wrapExpressionContent = (content: string): string => {
+  return (
+    `
+  (() => {
+      if (typeof ${content} === 'function') {
+          return ${content}
+      } else {
+          console.warn('${
+            errorMessages[ErrorCodes.X_V_ON_EXPRESSION_NOT_FUNCTION]
```

Comment:
`ErrorCodes` here are for compiler errors, and it doesn't make sense to use it in a runtime warnings, especially in generated code.

We should also avoid generating verbose code like this since it affects user bundle size.

Reply:
Thanks.